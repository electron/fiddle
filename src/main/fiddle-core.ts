import { ChildProcess } from 'node:child_process';

import { ElectronVersions, Installer, Runner } from '@electron/fiddle-core';
import { BrowserWindow, IpcMainEvent, WebContents } from 'electron';

import { ELECTRON_DOWNLOAD_PATH, ELECTRON_INSTALL_PATH } from './constants';
import { ipcMainManager } from './ipc';
import {
  DownloadVersionParams,
  ProgressObject,
  StartFiddleParams,
} from '../interfaces';
import { IpcEvents } from '../ipc-events';

let installer: Installer;
let runner: Runner;

// Keep track of which fiddle process belongs to which WebContents
const fiddleProcesses = new WeakMap<WebContents, ChildProcess>();

const downloadingVersions = new Map<string, Promise<any>>();
const removingVersions = new Map<string, Promise<void>>();

/**
 * Start running an Electron fiddle.
 */
export async function startFiddle(
  webContents: WebContents,
  params: StartFiddleParams,
): Promise<void> {
  const { dir, env, isValidBuild, localPath, options, version } = params;
  const child = await runner.spawn(
    isValidBuild && localPath ? Installer.getExecPath(localPath) : version,
    dir,
    { args: options, cwd: dir, env },
  );
  fiddleProcesses.set(webContents, child);

  const pushOutput = (data: string | Buffer) => {
    ipcMainManager.send(
      IpcEvents.FIDDLE_RUNNER_OUTPUT,
      [data.toString()],
      webContents,
    );
  };

  child.stdout?.on('data', pushOutput);
  child.stderr?.on('data', pushOutput);

  child.on('close', async (code, signal) => {
    fiddleProcesses.delete(webContents);

    ipcMainManager.send(IpcEvents.FIDDLE_STOPPED, [code, signal], webContents);
  });
}

/**
 * Stop a currently running Electron fiddle.
 */
export function stopFiddle(webContents: WebContents): void {
  const child = fiddleProcesses.get(webContents);
  child?.kill();

  if (child) {
    // If the child process is still alive 1 second after we've
    // attempted to kill it by normal means, kill it forcefully.
    setTimeout(() => {
      if (child.exitCode === null) {
        child.kill('SIGKILL');
      }
    }, 1000);
  }
}

export async function setupFiddleCore(versions: ElectronVersions) {
  // For managing downloads and versions for electron
  installer = new Installer({
    electronDownloads: ELECTRON_DOWNLOAD_PATH,
    electronInstall: ELECTRON_INSTALL_PATH,
  });

  // Broadcast state changes to all windows
  installer.on('state-changed', (event) => {
    for (const window of BrowserWindow.getAllWindows()) {
      ipcMainManager.send(
        IpcEvents.VERSION_STATE_CHANGED,
        [event],
        window.webContents,
      );
    }
  });

  runner = await Runner.create({ installer, versions });

  ipcMainManager.on(
    IpcEvents.GET_VERSION_STATE,
    (event: IpcMainEvent, version: string) => {
      event.returnValue = installer.state(version);
    },
  );
  ipcMainManager.handle(
    IpcEvents.DOWNLOAD_VERSION,
    async (
      event: IpcMainEvent,
      version: string,
      opts?: Partial<DownloadVersionParams>,
    ) => {
      const webContents = event.sender;

      if (removingVersions.has(version)) {
        throw new Error('Version is being removed');
      }

      if (!downloadingVersions.has(version)) {
        const promise = installer.ensureDownloaded(version, {
          ...opts,
          progressCallback: (progress: ProgressObject) => {
            ipcMainManager.send(
              IpcEvents.VERSION_DOWNLOAD_PROGRESS,
              [version, progress],
              webContents,
            );
          },
        });

        downloadingVersions.set(version, promise);
      }

      try {
        await downloadingVersions.get(version);
      } finally {
        downloadingVersions.delete(version);
      }
    },
  );
  ipcMainManager.handle(
    IpcEvents.REMOVE_VERSION,
    async (_: IpcMainEvent, version: string) => {
      if (downloadingVersions.has(version)) {
        throw new Error('Version is being downloaded');
      }

      if (!removingVersions.has(version)) {
        removingVersions.set(version, installer.remove(version));
      }

      try {
        await removingVersions.get(version);
        return installer.state(version);
      } finally {
        removingVersions.delete(version);
      }
    },
  );
  ipcMainManager.handle(
    IpcEvents.START_FIDDLE,
    async (event: IpcMainEvent, params: StartFiddleParams) => {
      await startFiddle(event.sender, params);
    },
  );
  ipcMainManager.on(IpcEvents.STOP_FIDDLE, (event: IpcMainEvent) => {
    stopFiddle(event.sender);
  });
}
