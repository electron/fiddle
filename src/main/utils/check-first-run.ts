import * as path from 'node:path';

import { app } from 'electron';
import * as fs from 'fs-extra';

const getConfigPath = () => {
  const userDataPath = app.getPath('userData');
  return path.join(userDataPath, 'FirstRun', 'electron-app-first-run');
};

/**
 * Whether or not the app is being run for
 * the first time
 */
export function isFirstRun(): boolean {
  const configPath = getConfigPath();

  try {
    if (fs.existsSync(configPath)) {
      return false;
    }

    fs.outputFileSync(configPath, '');
  } catch (error) {
    console.warn(`First run: Unable to write firstRun file`, error);
  }

  return true;
}
