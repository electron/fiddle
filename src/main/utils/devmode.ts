/**
 * Are we currently running in development mode?
 *
 * @returns {boolean}
 */
export function isDevMode(): boolean {
  return !!process.defaultApp;
}
