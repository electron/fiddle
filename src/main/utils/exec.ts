import { exec as cp_exec } from 'child_process';
import { promisify } from 'util';

import shellEnv from 'shell-env';

/**
 * On macOS & Linux, we need to fix the $PATH environment variable
 * so that we can call `npm`.
 *
 * @returns {Promise<void>}
 */
export const maybeFixPath = (() => {
  // Singleton: We don't want to do this more than once.
  let _shellPathCalled = false;

  return async (): Promise<void> => {
    if (_shellPathCalled) {
      return;
    }

    if (process.platform !== 'win32') {
      const { PATH } = await shellEnv();
      if (PATH) {
        process.env.PATH = PATH;
      }
    }

    _shellPathCalled = true;
  };
})();

/**
 * Execute a command in a directory.
 *
 * @param {string} dir
 * @param {string} cliArgs
 * @returns {Promise<string>}
 */
export async function exec(dir: string, cliArgs: string): Promise<string> {
  await maybeFixPath();

  const { stdout } = await promisify(cp_exec)(cliArgs, {
    cwd: dir,
    maxBuffer: 200 * 1024 * 100, // 100 times the default
  });

  return stdout.trim();
}
