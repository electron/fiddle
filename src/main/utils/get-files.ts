import { MessageChannelMain } from 'electron';

import { FileTransformOperation, Files } from '../../interfaces';
import { IpcEvents } from '../../ipc-events';
import { ipcMainManager } from '../ipc';

/**
 * Gets file content from the renderer
 */
export function getFiles(
  transforms: Array<FileTransformOperation>,
): Promise<{ localPath?: string; files: Files }> {
  return new Promise((resolve) => {
    const { port1, port2 } = new MessageChannelMain();
    ipcMainManager.postMessage(
      IpcEvents.GET_FILES,
      { options: undefined, transforms },
      [port1],
    );
    port2.once('message', (event) => {
      resolve(event.data);
      port2.close();
    });
    port2.start();
  });
}
