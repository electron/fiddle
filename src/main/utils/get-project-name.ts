import * as path from 'node:path';

import * as namor from 'namor';

/**
 * Returns a name for this project
 */
export function getProjectName(localPath?: string): string {
  if (localPath) {
    return path.basename(localPath);
  }

  return namor.generate({ words: 3, numbers: 0 });
}
