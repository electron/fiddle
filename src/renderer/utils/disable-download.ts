import semver from 'semver';

/**
 * disables download button for versions:
 * - below 11.0.0 on Apple Silicon.
 * - below 6.0.8 and 7.0.0 on Windows arm64
 * Reference: {@link https://www.electronjs.org/blog/apple-silicon}
 *
 * @param version - electron version
 */
export function disableDownload(version: string): boolean {
  return (
    (window.ElectronFiddle.platform === 'darwin' &&
      window.ElectronFiddle.arch === 'arm64' &&
      semver.lt(version, '11.0.0')) ||
    (window.ElectronFiddle.platform === 'win32' &&
      window.ElectronFiddle.arch === 'arm64' &&
      !semver.satisfies(version, '>=6.0.8 || >=7.0.0'))
  );
}
