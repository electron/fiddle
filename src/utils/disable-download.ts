import semver from 'semver';

/**
 * disables download button for versions:
 * - below 11.0.0 on Apple Silicon.
 * - below 6.0.8 and 7.0.0 on Windows arm64
 * Reference: {@link https://www.electronjs.org/blog/apple-silicon}
 *
 * @param {string} version - electron version
 * @returns {boolean}
 */
export function disableDownload(version: string) {
  return (
    (process.platform === 'darwin' &&
      process.arch === 'arm64' &&
      semver.lt(version, '11.0.0')) ||
    (process.platform === 'win32' &&
      process.arch === 'arm64' &&
      !semver.satisfies(version, '>=6.0.8 || >=7.0.0'))
  );
}
