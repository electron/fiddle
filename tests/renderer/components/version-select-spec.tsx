import * as React from 'react';

import { InstallState } from '@vertedinde/fiddle-core';
import { shallow } from 'enzyme';

import {
  ElectronReleaseChannel,
  RunnableVersion,
  VersionSource,
} from '../../../src/interfaces';
import {
  VersionSelect,
  filterItems,
  getItemIcon,
  getItemLabel,
  renderItem,
} from '../../../src/renderer/components/version-select';
import { disableDownload } from '../../../src/utils/disable-download';
import { StateMock, VersionsMock } from '../../mocks/mocks';

const { downloading, installed, missing, installing } = InstallState;
const { remote, local } = VersionSource;

jest.mock('../../../src/utils/disable-download.ts');

describe('VersionSelect component', () => {
  let store: StateMock;

  const mockVersion1 = {
    source: remote,
    state: missing,
    version: '1.0.0',
  };

  const mockVersion2 = {
    source: remote,
    state: missing,
    version: '3.0.0-unsupported',
  };

  beforeEach(() => {
    ({ state: store } = (window as any).ElectronFiddle.app);

    const { mockVersions } = new VersionsMock();
    store.initVersions('2.0.2', {
      ...mockVersions,
      '1.0.0': { ...mockVersion1 },
      '3.0.0-unsupported': { ...mockVersion2 },
    });
    store.channelsToShow = [
      ElectronReleaseChannel.stable,
      ElectronReleaseChannel.beta,
    ];
  });

  const onVersionSelect = () => ({});

  it('renders', () => {
    const wrapper = shallow(
      <VersionSelect
        appState={store as any}
        currentVersion={mockVersion1}
        onVersionSelect={onVersionSelect}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  describe('renderItem()', () => {
    it('renders an item', () => {
      const item = renderItem(mockVersion1, {
        handleClick: () => ({}),
        index: 0,
        modifiers: { active: true, disabled: false, matchesPredicate: true },
        query: '',
      });

      expect(item).toMatchSnapshot();
    });

    it('returns null if it does not match predicate', () => {
      const item = renderItem(mockVersion1, {
        handleClick: () => ({}),
        index: 0,
        modifiers: { active: true, disabled: false, matchesPredicate: false },
        query: '',
      });

      expect(item).toBe(null);
    });
  });

  describe('disableDownload', () => {
    it('disables download buttons when return value is true', () => {
      (disableDownload as any).mockReturnValueOnce(true);

      const item = renderItem(mockVersion1, {
        handleClick: () => ({}),
        index: 0,
        modifiers: { active: true, disabled: false, matchesPredicate: true },
        query: '',
      })!;

      const ItemWrapper = shallow(item);

      expect(ItemWrapper.find('.disabled-menu-tooltip')).toHaveLength(1);
    });

    it('does not disable enabled download buttons when return value is false', () => {
      (disableDownload as any).mockReturnValueOnce(false);

      const item = renderItem(mockVersion1, {
        handleClick: () => ({}),
        index: 0,
        modifiers: { active: true, disabled: false, matchesPredicate: true },
        query: '',
      })!;

      const ItemWrapper = shallow(item);

      expect(ItemWrapper.exists('.disabled-menu-tooltip')).toBe(false);
    });
  });

  describe('getItemLabel()', () => {
    it('returns the correct label for a local version', () => {
      const input: RunnableVersion = {
        ...mockVersion1,
        source: local,
      };

      expect(getItemLabel(input)).toBe('Local');
      expect(getItemLabel({ ...input, name: 'Hi' })).toBe('Hi');
    });

    it('returns the correct label for a version not downloaded', () => {
      const input: RunnableVersion = {
        ...mockVersion1,
        state: missing,
      };

      expect(getItemLabel(input)).toBe('Not downloaded');
    });

    it('returns the correct label for a version downloaded', () => {
      const input: RunnableVersion = {
        ...mockVersion1,
        state: installed,
      };

      expect(getItemLabel(input)).toBe('Downloaded');
    });

    it('returns the correct label for a version downloading', () => {
      const input: RunnableVersion = {
        ...mockVersion1,
        state: downloading,
      };

      expect(getItemLabel(input)).toBe('Downloading');
    });
  });

  describe('getItemIcon()', () => {
    it('returns the correct icon', () => {
      const icons: Array<{ state: InstallState; expected: string }> = [
        { state: downloading, expected: 'cloud-download' },
        { state: installed, expected: 'saved' },
        { state: missing, expected: 'cloud' },
        { state: installing, expected: 'compressed' },
      ];
      icons.forEach(({ state, expected }) => {
        expect(getItemIcon({ ...mockVersion1, state })).toBe(expected);
      });
    });
  });

  describe('filterItems()', () => {
    it('correctly matches a numeric query', () => {
      const versions = [
        { version: '14.3.0' },
        { version: '3.0.0' },
        { version: '13.2.0' },
        { version: '12.0.0-nightly.20210301' },
        { version: '12.0.0-beta.3' },
      ] as RunnableVersion[];

      const expected = [
        { version: '3.0.0' },
        { version: '13.2.0' },
        { version: '14.3.0' },
        { version: '12.0.0-beta.3' },
        { version: '12.0.0-nightly.20210301' },
      ] as RunnableVersion[];

      expect(filterItems('3', versions)).toEqual(expected);
    });

    it('sorts in descending order when the query is non-numeric', () => {
      const versions = [
        { version: '3.0.0' },
        { version: '13.2.0' },
        { version: '14.3.0' },
        { version: '12.0.0-beta.3' },
        { version: '3.0.0-nightly.12345678' },
        { version: '13.2.0-nightly.12345678' },
        { version: '14.3.0-nightly.12345678' },
        { version: '9.0.0-nightly.12345678' },
        { version: '12.0.0-nightly.20210301' },
      ] as RunnableVersion[];

      const expected = [
        { version: '14.3.0-nightly.12345678' },
        { version: '13.2.0-nightly.12345678' },
        { version: '12.0.0-nightly.20210301' },
        { version: '9.0.0-nightly.12345678' },
        { version: '3.0.0-nightly.12345678' },
      ] as RunnableVersion[];

      expect(filterItems('nightly', versions)).toEqual(expected);
    });
  });
});
