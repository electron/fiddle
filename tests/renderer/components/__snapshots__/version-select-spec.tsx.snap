// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VersionSelect component renderItem() renders an item 1`] = `
<Blueprint3.MenuItem
  active={true}
  disabled={false}
  icon="cloud"
  label="Not downloaded"
  multiline={false}
  onClick={[Function]}
  popoverProps={Object {}}
  shouldDismissPopover={true}
  text={
    Array [
      "1.0.0",
    ]
  }
/>
`;

exports[`VersionSelect component renders 1`] = `
<Blueprint3.Select
  disabled={false}
  filterable={true}
  itemListPredicate={[Function]}
  itemListRenderer={[Function]}
  itemRenderer={[Function]}
  items={
    Array [
      Object {
        "source": "remote",
        "state": "installed",
        "version": "2.0.3",
      },
      Object {
        "source": "remote",
        "state": "installed",
        "version": "2.0.2",
      },
      Object {
        "source": "remote",
        "state": "installed",
        "version": "2.0.1",
      },
      Object {
        "source": "remote",
        "state": "installed",
        "version": "2.0.0",
      },
      Object {
        "source": "remote",
        "state": "installed",
        "version": "1.8.7",
      },
      Object {
        "source": "remote",
        "state": "missing",
        "version": "1.0.0",
      },
      Object {
        "source": "remote",
        "state": "missing",
        "version": "3.0.0-unsupported",
      },
    ]
  }
  noResults={
    <Blueprint3.MenuItem
      disabled={true}
      multiline={false}
      popoverProps={Object {}}
      shouldDismissPopover={true}
      text="No results."
    />
  }
  onItemSelect={[Function]}
>
  <Blueprint3.Button
    className="version-chooser"
    disabled={false}
    icon="cloud"
    onContextMenu={[Function]}
    text="Electron v1.0.0"
  />
</Blueprint3.Select>
`;
